File: Python/Min-Heap/heap_recursive.py
class MinHeap:
    def __init__(self,capacity):
        self.storage = [0] * capacity
        self.capacity = capacity
        self.size = 0
    
    def getLeftChildIndex(self,index):
        return 2 * index + 1

    def getRightChildIndex(self,index):
        return 2 * index + 2

    def getParentIndex(self,index):
        return (index - 1) // 2

    def hasLeftChild(self,index):
        return self.getLeftChildIndex(index) < self.size

    def hasRightChild(self,index):
        return self.getRightChildIndex(index) < self.size

    def hasParent(self,index):
        return self.getParentIndex(index) >= 0
    
    def leftChild(self,index):
        return self.storage[self.getLeftChildIndex(index)]
    
    def rightChild(self,index):
        return self.storage[self.getRightChildIndex(index)]
    
    def parent(self,index):
        return self.storage[self.getParentIndex(index)]
    
    def isFull(self):
        return self.size == self.capacity 

    def swap(self,index1,index2):
        temp = self.storage[index1]
        self.storage[index1] = self.storage[index2]
        self.storage[index2] = temp
    
    def removeMin(self):
        if(self.size == 0):
            raise("Empty Heap")
        data = self.storage[0]
        self.storage[0] = self.storage[self.size - 1]
        self.size -= 1
        self.heapifyDown(0)
        return data
    
    def heapifyDown(self,index):
        smallest = index
        if(self.hasLeftChild(index) and self.storage[smallest] > self.leftChild(index)):
            smallest = self.getLeftChildIndex(index)
        if(self.hasRightChild(index) and self.storage[smallest] > self.rightChild(index)):
            smallest = self.getRightChildIndex(index)
        if(smallest != index):
            self.swap(index,smallest)
            self.heapifyDown(smallest)
    
    def insert(self,data):
        if(self.isFull()):
            raise Exception("Heap is Full")
        self.storage[self.size] = data 
        self.size += 1
        self.heapifyUp(self.size - 1)

    def heapifyUp(self,index):
        if(self.hasParent(index) and self.parent(index) > self.storage[index]): 
            self.swap(index,self.getParentIndex(index))
            self.heapifyUp(self.getParentIndex(index))

File: JavaScript/Min-Heap/heap_recursive.js
class MinHeap{
    constructor(){
        this.storage = [];
        this.size = 0;
    }

    getLeftChildIndex(index){
        return 2 * index + 1;
    }

    getRightChildIndex(index){
        return 2 * index + 2;
    }
    
    getParentIndex(index){
        return Math.floor((index - 1) / 2)
    }

    hasLeftChild(index){
        return this.getLeftChildIndex(index) < this.size;
    }

    hasRightChild(index){
        return this.getRightChildIndex(index) < this.size;
    }

    hasParent(index){
        return this.getParentIndex(index) >= 0;
    }

    leftChild(index){
        return this.storage[this.getLeftChildIndex(index)];
    }

    rightChild(index){
        return this.storage[this.getRightChildIndex(index)];
    }

    parent(index){
        return this.storage[this.getParentIndex(index)];
    }

    isFull(){
        return this.size == this.capacity;
    }

    swap(index1,index2){
        let temp = this.storage[index1];
        this.storage[index1] = this.storage[index2];
        this.storage[index2] = temp;
    }

    removeMin(){
        if(this.size == 0)
            throw new Error("Empty Heap");
        let data = this.storage[0];
        this.storage[0] = this.storage[this.size - 1];
        this.size -= 1;
        this.heapifyDown(0);
        return data
    }

    heapifyDown(index){
        let smallest = index;
        if(this.hasLeftChild(index) && this.storage[smallest] > this.leftChild(index))
            smallest = this.getLeftChildIndex(index);
        if(this.hasRightChild(index) && this.storage[smallest] > this.rightChild(index))
            smallest = this.getRightChildIndex(index);
        if(smallest != index){
            this.swap(index,smallest);
            this.heapifyDown(smallest);
        }
    }   

    insert(data){
        this.storage[this.size] = data; 
        this.size += 1;
        this.heapifyUp(this.size - 1);
    }

    heapifyUp(index){
        if(this.hasParent(index) && this.parent(index) > this.storage[index]{ 
            this.swap(index,this.getParentIndex(index));
            this.heapifyUp(this.getParentIndex(index));
        }
    }
}